File: src/testbed/runner.cpp
/*
 * This file is part of UNISPHERE.
 *
 * Copyright (C) 2013 Jernej Kos <jernej@kos.mx>
 *
 * This library is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
#include "testbed/runner.h"
#include "testbed/cluster/node.h"
#include "testbed/cluster/master.h"
#include "testbed/cluster/slave.h"
#include "testbed/cluster/controller.h"
#include "testbed/exceptions.h"

#include <boost/program_options.hpp>

namespace po = boost::program_options;

namespace UniSphere {

namespace TestBed {

/**
 * Cluster role.
 */
enum class ClusterRole {
  /// Node is a master and coordinates the cluster
  Master,
  /// Node is a slave and performs the simulations
  Slave,
  /// Node is a controller and submits commands to master
  Controller
};

std::istream &operator>>(std::istream &is, ClusterRole &role)
{
  std::string token;
  is >> token;
  if (token == "master")
    role = ClusterRole::Master;
  else if (token == "slave")
    role = ClusterRole::Slave;
  else if (token == "controller")
    role = ClusterRole::Controller;
  else
    throw boost::program_options::invalid_option_value(token);
  return is;
}

class RunnerPrivate {
public:
  RunnerPrivate();

  int run(int argc, char **argv);
public:
  /// Cluster node instance
  ClusterNodePtr m_clusterNode;
};

RunnerPrivate::RunnerPrivate()
{
}

int RunnerPrivate::run(int argc, char **argv)
{
  // Setup program options
  po::options_description options;

  po::options_description baseOptions("Base options");
  baseOptions.add_options()
    ("help", "displays help information")
    ("cluster-role", po::value<ClusterRole>(), "cluster role (master, slave, controller)")
  ;
  options.add(baseOptions);

  // Parse options
  po::variables_map vm;
  try {
    auto globalParsed = po::command_line_parser(argc, argv).options(baseOptions).allow_unregistered().run();
    po::store(globalParsed, vm);
    po::notify(vm);

    // Handle cluster role option before all others as it determines which module will be used
    if (vm.count("cluster-role")) {
      switch (vm["cluster-role"].as<ClusterRole>()) {
        case ClusterRole::Master: m_clusterNode = ClusterNodePtr(new Master); break;
        case ClusterRole::Slave: m_clusterNode = ClusterNodePtr(new Slave); break;
        case ClusterRole::Controller: m_clusterNode = ClusterNodePtr(new Controller); break;
      }
    } else {
      std::cout << "ERROR: No --cluster-role specified!" << std::endl;
      std::cout << options << std::endl;
      return 1;
    }

    m_clusterNode->initialize(argc, argv, options);
    if (vm.count("help")) {
      std::cout << "UNISPHERE Testbed" << std::endl;
      std::cout << options << std::endl;
      return 1;
    }

    return m_clusterNode->start();
  } catch (ArgumentError &e) {
    std::cout << "ERROR: There is an error in your invocation arguments!" << std::endl;
    std::cout << "ERROR: " << e.message() << std::endl;
    std::cout << options << std::endl;
    return 1;
  } catch (boost::program_options::error &e) {
    std::cout << "ERROR: There is an error in your invocation arguments!" << std::endl;
    std::cout << "ERROR: " << e.what() << std::endl;
    std::cout << options << std::endl;
    return 1;
  }

  return 0;
}

Runner::Runner()
  : d(new RunnerPrivate)
{
}

int Runner::run(int argc, char **argv)
{
  return d->run(argc, argv);
}

}

}

File: tools/testbed/runner.py
#
# This file is part of UNISPHERE.
#
# Copyright (C) 2013 Jernej Kos <jernej@kos.mx>
#
# This library is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

from .catalog import catalog
from . import exceptions

import argparse
import fnmatch
import hashlib
import importlib
import logging
import logging.config
import os
import sys
import traceback

logger = logging.getLogger('testbed.runner')

class Runner(object):
  def run(self, settings):
    """
    Scenario runner entry point.
    """
    main_parser = argparse.ArgumentParser("runner")
    main_parser.add_argument('--runs', metavar='name', type=str, nargs='+',
                             help='limit to specific runs')
    args = main_parser.parse_args()

    # Setup logging to stderr
    logging.config.dictConfig(settings.LOGGING)
    logger.info("Testbed root: %s" % settings.TESTBED_ROOT)

    # Select cluster configuration
    cluster_cfg = settings.CLUSTERS[settings.CLUSTER]

    logger.info("Loading cluster runner '%s'..." % settings.CLUSTER)
    i = settings.CLUSTER.rfind('.')
    module, attr = settings.CLUSTER[:i], settings.CLUSTER[i + 1:]
    try:
      module = importlib.import_module(module)
      cluster = getattr(module, attr)(cluster_cfg, settings)
    except (ImportError, AttributeError):
      raise exceptions.ImproperlyConfigured("Error importing cluster runner '%s'!" % settings.CLUSTER)

    logger.info("Loading run catalog...")
    catalog.load(settings, graphs=False)

    # Generate unique run identifier so that we can be sure that all runs have the same version
    run_id = hashlib.md5(os.urandom(32)).hexdigest()[:5]

    logger.info("Executing all runs (run_id=%s)..." % run_id)
    for descriptor in catalog.runs():
      if args.runs:
        for pattern in args.runs:
          if fnmatch.fnmatch(descriptor.name, pattern):
            break
        else:
          logger.info("Skipping run '%s'." % descriptor.name)
          continue

      logger.info("Starting run '%s'" % descriptor.name)
      for key, value in descriptor.settings.items():
        logger.info("  [%s] = %s" % (key, value))

      try:
        descriptor.run(cluster, run_id)
      except KeyboardInterrupt:
        logger.info("Abort requested by user.")
        return
      except exceptions.ScenarioRunFailed:
        logger.error("Aborting due to scenario run error.")
        return
      except:
        logger.error("Aborting due to error.")
        logger.error(traceback.format_exc())
        return

      logger.info("Run '%s' completed." % descriptor.name)
